### PR TITLE
Grammar based EL expressions transpiler

### PR DESCRIPTION
**Pull request by nuclearpinguin**
Make sure you have checked _all_ steps below and that above description is correct.

### JIRA
- [ ] My PR addresses the following issue #179 

### Tests
- [ ] My PR has unit tests testing the functionality (or I explained why it needs no tests)

### Commits
- [ ] My commits reference the issue in the message using: Fixes #XXX /Closes #XXX or similar.

### Documentation
- [ ] My PR adds documentation that describes how to use it (or I explained why it needs no docs)

### CODE REVIEW

**Comment by nuclearpinguin** [comment]
I signed it!

**Comment by nuclearpinguin** [comment]
I signed it!

**Comment by nuclearpinguin** [comment]
In the `"${'${'}` edge case one can convert outer `$` to `#`: `"#{'${'}`.

File: o2a/o2a_libs/el_parser.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        | "!"
+        | "not"
+        | "empty"
+
+    value: value_prefix (value_suffix)*
+
+    value_prefix: literal
+        | non_litera_lvalue_prefix
+
+    non_litera_lvalue_prefix: "(" expression ")"
```
Comment (potiuk) [inline o2a/o2a_libs/el_parser.py]: should be non_literal_lvalue_prefix maybe ?

File: o2a/o2a_libs/el_parser.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+    value_suffix: "." identifier | "[" expression "]"
+
+    identifier: JAVA
+
+    function_invocation: (identifier ":")? identifier "(" ( expression ( "," expression )* )? ")"
+
+    literal: BOOL | INT | FLOAT | STRING | NULL
+
+    JAVA: /[a-zA-Z_]+/
```
Comment (potiuk) [inline o2a/o2a_libs/el_parser.py]: As discussed - this is not full regexp yet, but it is good enough for now.

File: o2a/o2a_libs/el_parser.py (hunk 1 above)
Comment (nuclearpinguin) [reply o2a/o2a_libs/el_parser.py]: You are right!

File: tests/o2a_libs/test_el_parser.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            ),
+            (
+                "${wf:actionData('getDirInfo')['dir.num-files'] gt 23 || "
+                "wf:actionData('getDirInfo')['dir.age'] gt 6}",
+                "{{wf_action_data('getDirInfo')['dir.num-files'] > 23 or "
+                "wf_action_data('getDirInfo')['dir.age'] > 6}}",
+            ),
+            ("#{'${'}", "{{'${'}}"),
+            ("${'${'}", "{{'${'}}"),
+            ("${function()} literal and #{OtherFunction}", "{{function()}} literal and{{other_function}}"),
```
Comment (nuclearpinguin) [inline tests/o2a_libs/test_el_parser.py]: I am not sure if this example is parsed correctly.  Should we skip the space before Jinja expression?
`...and{{..." -> "...and {{...`

File: o2a/o2a_libs/el_parser.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+__all__ = ["translate"]
+
+from typing import Union
+import re
+
+from lark import Lark, Tree, Token
+
+
+GRAMMAR = r"""
```
Comment (fenglu-g) [inline o2a/o2a_libs/el_parser.py]: Is this just a reproduction of the JSP lang spec? e.g., 

LValue ::= ‘${‘ LValueInner ‘}’
| ‘#{‘ LValueInner ‘}’
...

File: o2a/o2a_libs/el_parser.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    %ignore " "
+"""
+
+
+def _parser(sentence: str) -> Tree:
+    return Lark(GRAMMAR, start="start", keep_all_tokens=True).parse(sentence)
+
+
+def _camel_to_snake(name: str) -> str:
```
Comment (fenglu-g) [inline o2a/o2a_libs/el_parser.py]: are there standard py libs for this? seems like a common need.

File: o2a/o2a_libs/el_parser.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    if isinstance(tree, Token):
+        return _translate_token(tree)
+
+    if tree.data == "function_invocation":
+        tree = _translate_function(tree)
+
+    if tree.data == "binary_op":
+        tree = _translate_binary_operator(tree)
+
+    if tree.data == "expression" and "ternary" in [ch.data for ch in tree.children]:
```
Comment (fenglu-g) [inline o2a/o2a_libs/el_parser.py]: could you give me an example on the two ternary cases?

File: o2a/o2a_libs/el_parser.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        return _translate_ternary(tree, case=1)
```
Comment (fenglu-g) [inline o2a/o2a_libs/el_parser.py]: nit: case ={1,2} isn't very readable.

File: o2a/o2a_libs/el_parser.py (hunk 4 above)
Comment (nuclearpinguin) [reply o2a/o2a_libs/el_parser.py]: Exactly, those production are based on the specification. However, the grammar is changed a little bit to simplify it and avoid some recursion errors.

File: o2a/o2a_libs/el_parser.py (hunk 5 above)
Comment (nuclearpinguin) [reply o2a/o2a_libs/el_parser.py]: It seems that using regex is the most popular approach and probably the simplest.

Comment (potiuk) [reply o2a/o2a_libs/el_parser.py]: We all got curious about it but it seems this is THE most popular solution :) with 700 upvotes on Stack Overflow

File: o2a/o2a_libs/el_parser.py (hunk 6 above)
Comment (nuclearpinguin) [reply o2a/o2a_libs/el_parser.py]: The first case is:
`f() ? true : false` 
Screenshot 2019-06-27 at 11 23 44

second one is:
` x op y ? true : false` where `op` is binary operator`.

Screenshot 2019-06-27 at 11 24 41

File: o2a/o2a_libs/el_parser.py (hunk 7 above)
Comment (nuclearpinguin) [reply o2a/o2a_libs/el_parser.py]: Agreed! I think it would be best to move this logic to `_translate_ternary()`.

**Pull request merged**